src/classification/mondrian_node.rs

```diff
@@ -0,0 +1,179 @@
+
+use std::fmt;
+
+use std::ops::Add;
+
+use std::usize;
+
+/// Node struct
+#[derive(Clone)]
+pub struct Node<F> {
+    pub parent: Option<usize>,
+    pub time: F, // Time: how much I increased the size of the box
+    pub is_leaf: bool,
+    pub min_list: Array1<F>, // Lists representing the minimum and maximum values of the data points contained in the current node
+    pub max_list: Array1<F>,
+    pub feature: usize, // Feature in which a split occurs
+    pub threshold: F,   // Threshold in which the split occures
+    pub left: Option<usize>,
+    pub right: Option<usize>,
+    pub stats: Stats<F>,
+}
+impl<F: FType> Node<F> {
+    pub fn update_leaf(&mut self, x: &Array1<F>, y: usize) {
+        self.stats.add(x, y);
+    }
+    pub fn update_internal(&self, left_s: &Stats<F>, right_s: &Stats<F>) -> Stats<F> {
+        left_s.merge(right_s)
+    }
+    /// Check if all the labels are the same in the node.
+    /// e.g. y=2, stats.counts=[0, 1, 10] -> False
+    /// e.g. y=2, stats.counts=[0, 0, 10] -> True
+    /// e.g. y=1, stats.counts=[0, 0, 10] -> False
+    pub fn is_dirac(&self, y: usize) -> bool {
+        return self.stats.counts.sum() == self.stats.counts[y];
+    }
+}
+
+/// Stats assocociated to one node
+///
+/// In nel215 code it is "Classifier"
+#[derive(Clone)]
+pub struct Stats<F> {
+    pub sums: Array2<F>,
+    pub sq_sums: Array2<F>,
+    pub counts: Array1<usize>,
+    n_labels: usize,
+}
+impl<F: FType + fmt::Display> fmt::Display for Stats<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "\n┌ Stats")?;
+        // sums
+        write!(f, "│ sums: [")?;
+        for row in self.sums.outer_iter() {
+            write!(f, "{:?}, ", row.to_vec())?;
+        }
+        writeln!(f, "]")?;
+        // sq_sums
+        write!(f, "│ sq_sums: [")?;
+        for row in self.sq_sums.outer_iter() {
+            write!(f, "{:?}, ", row.to_vec())?;
+        }
+        writeln!(f, "]")?;
+        // count
+        write!(f, "└ counts: {}", self.counts)?;
+        Ok(())
+    }
+}
+impl<F: FType> Stats<F> {
+    pub fn new(n_labels: usize, n_features: usize) -> Self {
+        Stats {
+            sums: Array2::zeros((n_labels, n_features)),
+            sq_sums: Array2::zeros((n_labels, n_features)),
+            counts: Array1::zeros(n_labels),
+            n_labels,
+        }
+    }
+    pub fn create_result(&self, x: &Array1<F>, w: F) -> Array1<F> {
+        let probs = self.predict_proba(x);
+        probs * w
+    }
+    pub fn add(&mut self, x: &Array1<F>, y: usize) {
+        // Same as: self.sums[y] += x;
+        self.sums.row_mut(y).zip_mut_with(&x, |a, &b| *a += b);
+
+        // Same as: self.sq_sums[y] += x*x;
+        // e.g. x: [1.059 0.580] -> x*x: [1.122  0.337]
+        self.sq_sums
+            .row_mut(y)
+            .zip_mut_with(&x, |a, &b| *a += b * b);
+
+        self.counts[y] += 1;
+    }
+    fn merge(&self, s: &Stats<F>) -> Stats<F> {
```

Comment:
Something to consider in the future: this seems to be an incremental variance estimator, in fact, the so-called naive estimator. This estimator is prone to numerical instabilities, namely, numerical cancellation.

In the main River repo, we made all tree estimators relying on Welford's algorithm. This estimator also has merging capabilities and also uses a similar memory footprint. On the other hand, it is way more stable.

Numerical cancellation might not be the main culprit behind the performance drop you are experiencing, @MarcoDiFrancesco, but it could be an additional factor.

In these [slides](https://smastelini.github.io/pdf/qo-waikato.pdf) I discussed the issues with the naive incremental variance estimator and how to solve them.

Reply:
That's a good point! We could implement it in the future.
I doubt numerical cancellation is the issue behind the performance drop. I tested everthing with _f64_ instead of _f32_ and the result is practically the same.
Nevertheless, as you mention in the slides it's definitely useful when dealing with large datasets so we should implement it later.